### PR TITLE
Js debugger react

### PR DESCRIPTION
**Pull request by Bjvanminnen**
I started doing some actual run mode changes (making the code workspace header color change while running). In doing that, I started moving some styling out of scss and into the component. There was some confusion because we have some of the same styles used in codeWorkspace (which is React) and in js debugger (which is ejs backed). I decided this justified moving the JS Debugger into React, which actually makes a few things cleaner.

The change should all just be cleanup, with no actual difference to the UI.

### CODE REVIEW

File: apps/src/applab/applab.js [hunk 1]
```diff
@@ -663,11 +663,6 @@ Applab.init = function (config) {
 
   if (showDebugButtons || showDebugConsole) {
     debuggerUi = new JsDebuggerUi(Applab.runButtonClick);
-    var extraControlRowsHtml = debuggerUi.getMarkup(studioApp.assetUrl, {
-      showButtons: showDebugButtons,
-      showConsole: showDebugConsole
-    });
-    extraControlRows = <ProtectedStatefulDiv dangerouslySetInnerHTML={{ __html : extraControlRowsHtml }} />;
```
Comment (Bjvanminnen) [inline apps/src/applab/applab.js:671]: No more using extraControlRows for what was actually JsDebugger :)

File: apps/src/applab/applab.js [hunk 2]
```diff
@@ -855,7 +853,7 @@ Applab.init = function (config) {
       localeDirection={studioApp.localeDirection()}
       editCode={!!config.level.editCode}
       readonlyWorkspace={Applab.reduxStore.getState().level.isReadOnlyWorkspace}
-      extraControlRows={extraControlRows}
+      showDebugger={showDebugButtons || showDebugConsole}
```
Comment (Bjvanminnen) [inline apps/src/applab/applab.js:856]: Another option instead of passing this in would be to always render JsDebugger, and have JsDebugger render an empty div in the case it's not really wanted. Not sure which is better. Thoughts?

File: apps/src/applab/reducers.js [hunk 3]
```diff
@@ -48,7 +48,10 @@ function level(state, action) {
         'instructionsMarkdown',
         'instructionsInTopPane',
         'puzzleNumber',
-        'stageTotal'
+        'stageTotal',
+        'showDebugButtons',
+        'showDebugConsole',
+        'showDebugWatch',
```
Comment (Bjvanminnen) [inline apps/src/applab/reducers.js:54]: The initialLevelProps reducers in applab and gamelab have a lot in common. Wonder if we want to share that code in some way eventually.

File: apps/src/templates/JsDebugger.jsx [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+var DebugWatch = function (props) {
+  var classes = 'debug-watch';
+  if (!props.debugButtons) {
+    classes += 'no-commands';
+  }
+  return (
+    <div id="debug-watch" className={classes}/>
+  );
+};
```
Comment (Bjvanminnen) [inline apps/src/templates/JsDebugger.jsx:89]: Thoughts on splitting these three little bits out into their own components in the same file like this?

File: apps/src/JsDebuggerUi.js [hunk 5]
```diff
@@ -9,6 +9,7 @@ var JSInterpreter = require('./JSInterpreter');
 var Observer = require('./Observer');
 var Slider = require('./slider');
 var utils = require('./utils');
+var JsDebugger = require('./templates/JsDebugger');
```
Comment (islemaster) [inline apps/src/JsDebuggerUi.js]: Unused?

File: apps/src/applab/reducers.js (hunk 3 above)
Comment (islemaster) [inline apps/src/applab/reducers.js:54]: initialLevelProps duck module?

File: apps/src/applab/applab.js [hunk 6]
```diff
@@ -36,6 +36,7 @@ var applabCommands = require('./commands');
 var JSInterpreter = require('../JSInterpreter');
 var JsInterpreterLogger = require('../JsInterpreterLogger');
 var JsDebuggerUi = require('../JsDebuggerUi');
+var JsDebugger = require('../templates/JsDebugger');
```
Comment (islemaster) [inline apps/src/applab/applab.js]: Unused?

File: apps/src/gamelab/GameLab.js [hunk 7]
```diff
@@ -14,6 +14,7 @@ var dropletUtils = require('../dropletUtils');
 var _ = require('../lodash');
 var dropletConfig = require('./dropletConfig');
 var JsDebuggerUi = require('../JsDebuggerUi');
+var JsDebugger = require('../templates/JsDebugger');
```
Comment (islemaster) [inline apps/src/gamelab/GameLab.js]: Unused?

File: apps/src/applab/applab.js (hunk 2 above)
Comment (islemaster) [inline apps/src/applab/applab.js:856]: I'm not sure one is better, but since we're already passing these properties into redux, I'd go ahead and let `JsDebugger` own this logic and render nothing if they're disabled, since it's already a connected component.

FYI, now that we're on React 15+ you can render `null` instead of an empty div.  [React will render it as an HTML comment.](https://facebook.github.io/react/blog/2016/04/07/react-v15.html#rendering-null-now-uses-comment-nodes)

Comment (islemaster) [inline apps/src/applab/applab.js:856]: Disregard: As long as we're using `JsDebugger` in the non-connected `CodeWorkspace` context, we need to do things the way you've got it now.

File: apps/src/applab/reducers.js (hunk 3 above)
Comment (Bjvanminnen) [inline apps/src/applab/reducers.js:54]: I think that might be the way to go. If so, belongs in a separate PR though.

File: apps/src/templates/JsDebugger.jsx (hunk 4 above)
Comment (islemaster) [inline apps/src/templates/JsDebugger.jsx:89]: :+1: Would be nice to also break slider-cell into one of these, even if we don't put it in a new file _yet._

File: apps/src/templates/JsDebugger.jsx [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+        <button id="continueButton" className="debugger_button">
+          <img src="/blockly/media/1x1.gif" className="continue-btn icon21"/>
+          {i18n.continue()}
+        </button>
+        <span> </span>
+        <button id="stepOverButton" className="debugger_button">
+          <img src="/blockly/media/1x1.gif" className="step-over-btn icon21"/>
+          {i18n.stepOver()}
+        </button>
+
```
Comment (islemaster) [inline apps/src/templates/JsDebugger.jsx:63]: What's with all the extra `<span> </span>` in this markup?
- Why isn't there one between stepOver and stepOut? I assume because we expect a line-break there.
- Why _is_ there one before pauseButton and one between pauseButton and continueButton (which are never displayed at the same time)?
- Could you just use nonbreaking spaces for this?

**Comment by islemaster** [comment]
Still need to delete JsDebuggerUi.html.ejs?

File: apps/src/templates/JsDebugger.jsx (hunk 8 above)
Comment (Bjvanminnen) [inline apps/src/templates/JsDebugger.jsx:63]: The spans are to insert whitespace so that things behave identically to the EJS file. The way I've done that elsewhere is

```
{" " /* Explicitly insert whitespace so that this behaves like our ejs file*/}
```

Not too sure why I did it differently here. Prob better to do it the same way I've been doing it elsewhere.

**Comment by Bjvanminnen** [comment]
@islemaster Thanks for all the great reviews! I think I've addressed all your comments here, and it's ready for another look.

File: apps/src/templates/JsDebugger.jsx [hunk 9]
```diff
@@ -89,12 +89,58 @@ var DebugWatch = function (props) {
 };
 
 /**
+ * Slider for modifying speed. Ideall we will eventually have a common component
```
Comment (islemaster) [inline apps/src/templates/JsDebugger.jsx:92]: s/Ideall/Ideally

File: apps/src/templates/JsDebugger.jsx [hunk 10]
```diff
@@ -45,17 +45,17 @@ var DebugButtons = function () {
...
       <div id="debug-buttons">
-        <span> </span>
+        {" " /* Explicitly insert whitespace so that this behaves like our ejs file*/}
         <button id="pauseButton" className="debugger_button">
           <img src="/blockly/media/1x1.gif" className="pause-btn icon21"/>
           {i18n.pause()}
         </button>
-        <span> </span>
+        {" " /* Explicitly insert whitespace so that this behaves like our ejs file*/}
         <button id="continueButton" className="debugger_button">
```
Comment (islemaster) [inline apps/src/templates/JsDebugger.jsx:54]: I appreciate preserving EJS behavior exactly, but I suspect the intended behavior here is pretty clear... which is why I question the space between `pauseButton` and `continueButton`.  Fine for now, but I'd feel free to rework these extra spaces whenever we manage to move the business logic into this component too.

File: apps/src/templates/JsDebugger.jsx [hunk 11]
```diff
@@ -115,43 +161,7 @@ var JsDebugger = function (props) {
...
-            <image
-                xlinkHref="/blockly/media/turtle_icons.png"
-                height={42}
-                width={84}
-                x={120}
-                y={-19}
-                clipPath="url(#fastClipPath)" />
-          </svg>
-        </div>
+        <Slider style={sliderStyle}/>
```
Comment (islemaster) [inline apps/src/templates/JsDebugger.jsx:164]: :+1:

**Comment by islemaster** [comment]
LGTM with new changes.  Thanks for doing this, I'm thrilled to see the debugger go React!  I bet we're _very_ close to being able to move most of its important behaviors into React too, so it's no longer "protected" DOM.

File: apps/src/templates/JsDebugger.jsx (hunk 10 above)
Comment (Bjvanminnen) [inline apps/src/templates/JsDebugger.jsx:54]: Yeah, in a lot of cases it's probably fine to remove, but it will result in a slightly different render. If we're going to do that, I'd like to do it in a separate pass (i.e. pass 1 is lots of code change, but eyes tests will pass because things render identically, pass 2 is minimal code change with eyes tests failing because things render slightly differently).
